File: java/ToolGood/Words/WordsSearch.java
package ToolGood.Words;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

import org.springframework.util.StringUtils;

import ToolGood.Words.internals.TrieNode2;

public class WordsSearch{
    protected TrieNode2[] _first = new TrieNode2[Character.MAX_VALUE + 1];

    public void SetKeywords(List<String> keywords)
    {
        Map<String, Integer> dict = new Hashtable<String, Integer>();
        for (int i = 0; i < keywords.size(); i++) {
            String item=keywords.get(i);
            dict.put(item, i);
        }
        SetKeywords(dict);
    }

    public void SetKeywords(Map<String,Integer> _keywords)
    {
        TrieNode2[] first =  new TrieNode2[Character.MAX_VALUE + 1];
        TrieNode2 root = new TrieNode2();

        _keywords.forEach((p,index)->{
            if(StringUtils.isEmpty(p)==false){ 
                TrieNode2 nd = _first[ p.charAt(0)];
                if (nd == null) {
                    nd = root.Add(p.charAt(0));
                    first[p.charAt(0)] = nd;
                }
                for (int i = 1; i < p.length(); i++) {
                    nd = nd.Add(p.charAt(i));
                }
                nd.SetResults(p,index);
            }
        });
 
        this._first = first;// root.ToArray();

        Map<TrieNode2, TrieNode2> links = new Hashtable<TrieNode2, TrieNode2>();
        root.m_values.forEach((k,v)->{
            TryLinks(v, null, links);
        });
        links.forEach((k,v)->{
            k.Merge(v, links);
        });
    }

    private void TryLinks(TrieNode2 node, TrieNode2 node2, Map<TrieNode2, TrieNode2> links)
    {
        node.m_values.forEach((Key,Value)->{
            TrieNode2 tn = null;
            if (node2 == null) {
                tn = _first[Key];
                if (tn != null) {
                    links.put(Value, tn);
                }
            } else if (node2.TryGetValue(Key, tn)) {
                links.put(Value, tn);
            }
            TryLinks(Value, tn, links);
        });
    }



    public WordsSearchResult FindFirst(String text)
    {
        TrieNode2 ptr = null;
        for (int i = 0; i < text.length(); i++) {
            Character t=text.charAt(i);
            TrieNode2 tn=null;
            if (ptr == null) {
                tn = _first[t];
            } else {
                if (ptr.TryGetValue(t, tn) == false) {
                    tn = _first[t];
                }
            }
            if (tn != null) {
                if (tn.End) {
                    for( String  key : tn.Results.keySet()){
                        Integer value= tn.Results.get(key);
                        return new WordsSearchResult(key, i + 1 - key.length(), i, value);
                    }
                }
            }
            ptr = tn;
        }
        return null;
    }

    public List<WordsSearchResult> FindAll(String text)
    {
        TrieNode2 ptr = null;
        List<WordsSearchResult> list = new ArrayList<WordsSearchResult>();

        for (int i = 0; i < text.length(); i++) {
            Character t=text.charAt(i);
            TrieNode2 tn=null;
            if (ptr == null) {
                tn = _first[t];
            } else {
                if (ptr.TryGetValue(t, tn) == false) {
                    tn = _first[t];
                }
            }
            if (tn != null) {
                if (tn.End) {
                    for( String  key : tn.Results.keySet()){
                        Integer value= tn.Results.get(key);
                        WordsSearchResult item= new WordsSearchResult(key, i + 1 - key.length(), i, value);
                        list.add(item);
                    }
                }
            }
            ptr = tn;
        }
        return list;
    }





    public boolean ContainsAny(String text)
    {
        TrieNode2 ptr = null;
        for (int i = 0; i < text.length(); i++) {
            Character t=text.charAt(i);
            TrieNode2 tn=null;
            if (ptr == null) {
                tn = _first[t];
            } else {
                if (ptr.TryGetValue(t, tn) == false) {
                    tn = _first[t];
                }
            }
            if (tn != null) {
                if (tn.End) {
                    return true;
                }
            }
            ptr = tn;
        }
        return false;
    }

    public String Replace(String text){
        return Replace(text,'*');
    }

    public String Replace(String text,Character replaceChar)
    {
        StringBuilder result = new StringBuilder(text);

        TrieNode2 ptr = null;
        for (int i = 0; i < text.length(); i++) {
            Character t=text.charAt(i);
            TrieNode2 tn=null;
            if (ptr == null) {
                tn = _first[t];
            } else {
                if (ptr.TryGetValue(t,   tn) == false) {
                    tn = _first[t];
                }
            }
            if (tn != null) {
                if (tn.End) {
                    Integer maxLength=0;
                    for(String key : tn.Results.keySet()){
                         if(key.length()>maxLength){
                            maxLength = key.length();
                        }
                    }
                    int start = i + 1 - maxLength;
                    for (int j = start; j <= i; j++) {
                        result.setCharAt(j, replaceChar);
                    }
                }
            }
            ptr = tn;
        }
        return result.toString();
    }

}

File: golang/ToolGood/Words/WordsSearch.go
package Words

import . "./internals"


type WordsSearch struct {
	_first []*TrieNode2
}

func NewWordsSearch() *WordsSearch  {
	return &WordsSearch{
		_first : make([]*TrieNode2,0),
	}
}

func (this *WordsSearch)SetKeywords(keywords []string){
	first := make([]*TrieNode2,0x10000)
	root := NewTrieNode2()
	for	i,p := range keywords {
		length:=len(p)
		if length>0 {
			var nd *TrieNode2
			for	i,ch := range p{
				if i==0 {
					nd = first[ch]
					if nd==nil {
						nd=root.Add(ch)
						first[ch] = nd;
					}
				}else{
					nd = nd.Add(ch);
				}
			}
			nd.SetResults(p,i);
		}
	}
	this._first=first
	links:= make(map[*TrieNode2]*TrieNode2)
	for _,val := range root.M_values {
		this.tryLinks(val, nil, links);
	}
	for key,val := range links {
		key.Merge(val,links)
 	}
}


func (this *WordsSearch) tryLinks(node *TrieNode2,node2 *TrieNode2,links map[*TrieNode2]*TrieNode2){
	for key,value:=range node.M_values {
		var tn *TrieNode2
		if node2 == nil {
			tn = this._first[key]
			if tn != nil {
				links[value]=tn
			}
		} else {
			b,tn:= node2.TryGetValue(key)
			if b==true{
				links[value]=tn
			}
		}
		this.tryLinks(value,tn,links)
	}
}


func (this *WordsSearch) FindFirst(text string) *WordsSearchResult{
	var ptr *TrieNode2
	var i int
	for _,t := range text {
		var tn *TrieNode2
		if ptr==nil{
			tn = this._first[t];
		}else{
			var b bool
			b,tn =ptr.TryGetValue(t)
			if b==false{
				tn = this._first[t];
			}
		}
		if tn!=nil {
			if tn.End==true {
				for k,v:= range tn.Results{
					maxLength:=len([]rune(k))
					return NewWordsSearchResult(k, i + 1 - maxLength, i, v);
				}
			}
		}
		ptr = tn
		i++
	}
	return nil
}



func (this *WordsSearch) FindAll(text string) []*WordsSearchResult{
	list := make([]*WordsSearchResult,0) 
	var ptr *TrieNode2
	var i int
	for _,t := range text {
		var tn *TrieNode2
		if ptr==nil{
			tn = this._first[t];
		}else{
			var b bool
			b,tn =ptr.TryGetValue(t)
			if b==false{
				tn = this._first[t];
			}
		}
		if tn!=nil {
			if tn.End==true {
				for k,v:= range tn.Results{
					maxLength:=len([]rune(k))
					r:= NewWordsSearchResult(k, i + 1 - maxLength, i, v);
					list=append(list,r)
				}
			}
		}
		ptr = tn;
		i++
	}
	return list
}


func (this *WordsSearch) ContainsAny(text string) bool{
	var ptr *TrieNode2
	for _,t := range text {
		var tn *TrieNode2
		if ptr==nil{
			tn = this._first[t];
		}else{
			var b bool
			b,tn =ptr.TryGetValue(t)
			if b==false{
				tn = this._first[t];
			}
		}
		if tn!=nil {
			if tn.End==true {
				return true
			}
		}
		ptr = tn;
	}
	return false
}


func (this *WordsSearch) Replace(text string,replaceChar int32) string{
	result:= []rune(text)
	var ptr *TrieNode2
	var i int
	for _,t := range text {
		var tn *TrieNode2
		if ptr==nil{
			tn = this._first[t];
		}else{
			var b bool
			b,tn =ptr.TryGetValue(t)
			if b==false{
				tn = this._first[t];
			}
		}
		if tn!=nil {
			if tn.End==true {
				for k,_:= range tn.Results{
					maxLength:=len([]rune(k))
					start:= i + 1 - maxLength
					for j := start; j <= i; j++{
						result[j]=replaceChar
					} 
					break
				}
			}
		}
		ptr = tn;
		i++
	}
	return string (result) 
}
